_data/open.yml

```diff
@@ -641,7 +641,7 @@ concepts:
     
     - concept: bell-polynomial
       arity: n
-      en: bell polynomial
+      en: $1 th bell polynomial of $2, $3, ...
```

Comment:
I don't know the definition of `n`. I assume n is a language construct indicating a variable argument length. However, this might cause problems when people try to parse the YML file and assume it would be a number. 

While the syntax in the notation field `$2, ... $n` seems unambiguous,  `$1 th bell polynomial of $2, $3, ...` is not easily translatable. For example, if we wanted to use translatewiki.net I would not know how to write the variable length argument https://www.mediawiki.org/wiki/Manual:Messages_API . The MediaWiki solution would be the arity is 2, and the second argument is a comma-separated list.

Reply:
Technically, one would also need a description for each label that discusses the arguments. Otherwise, it is hardly possible to translate. If we would expose it to translatewiki, we would get up to 300 language variants. 

See for example 

- [description](https://github.com/wikimedia/mediawiki-extensions-Math/blob/master/i18n/qqq.json#L82C1-L82C204
) 
```json
	"math_invalidresponse": "Used as error message.\n\nFollows the message {{msg-mw|Math failure}}.\n\nParameters:\n* $1 - rendering type (either LaTeXML or Mathoid)\n* $2 - hostname\n* $3 - error message",
```
- [en message](https://github.com/wikimedia/mediawiki-extensions-Math/blob/008db7ad08f3fc6b4d105d03791d6ff1fc858011/i18n/en.json#L82)
```json
	"math_invalidresponse": "$1: Invalid response (\"$3\") from server \"$2\":",
```
- [mk message](https://github.com/wikimedia/mediawiki-extensions-Math/blob/008db7ad08f3fc6b4d105d03791d6ff1fc858011/i18n/mk.json#L70)
```json
	"math_invalidresponse": "$1: Неважечки одговор („$3“) од опслужувачот „$2“:",
```

-  [zh-hant message](https://github.com/wikimedia/mediawiki-extensions-Math/blob/008db7ad08f3fc6b4d105d03791d6ff1fc858011/i18n/zh-hant.json#L84)

```json
	"math_invalidresponse": "$1：從伺服器 \"$2\" 收到無效的回應 (\"$3\")。",
```
- [hi message](https://github.com/wikimedia/mediawiki-extensions-Math/blob/008db7ad08f3fc6b4d105d03791d6ff1fc858011/i18n/hi.json#L58)
```json
	"math_invalidresponse": "सर्वर '$2' से अमान्य लेटेक्सएमएल उत्तर ('$3')",
```

While one can possible generate the description part of the qqq message automatically, it would be good to have an option to describe the parameters.

Reply:
@physikerwelt  ` don't know the definition of n.`

er yes.....

The data quality is currently somewhat low it was converted as a complete collection from multiple google docs that were not all in the same format and none of which (being google doc files) had any much usable structure. So Iteratively the edits are getting first to at least valid yaml with a consistent display, and then the actual entries need to be reviewed.

A large set of these, I really have no idea how to read. A practitioner in the area may well ignore the name and subscripts etc and just read a bell polynomial as "B n k"  for what wikipedia says is the "complete exponential bell polynomial"  perhaps in the end we will need to give separate verbose and terse templates (as just done in the speech templates for characters). Some of the notations here I have really no idea how I'd read them, even after looking up the examples in the suplied URL. i
I tried to make sure that (for example) an arity 3 concept has a default speech template of `this concept name $1 $2 $3` so it all gets read, but they all need changing or at least reviewing.

On translations: yes they will be needed, the format allows multiple translations using the relevant country or language code, it will loop through the provided ones making extra columns that can be displayed using the menu at the top, but I think we need to get the english ones in a better state before contemplating inviting translations.

In core we have got away with informally describing the argument meaning for all arguments together as part of a single comment.  Are you suggesting that the yaml should have specific field for each argument that describes it?

so something like


```
    - concept: fraction
      arity: 2
      property: function
      en: fraction $1 over $2 end fraction
```

becoming

```
    - concept: fraction
      arity: 2
      arguments:
        - numerator
        - denominator
      property: function
      en: fraction $1 over $2 end fraction
```

?

to allow a more structured argument description?

It would be possible of course, probably best to make it optional as adding them to all entries all at once as a mandatory change in the format might be a bit tricky.

Reply:
The mandatory arguments in the style of `this concept name $1 $2 $3` make translation more manual. If we didn't have the arguments, we could use the translations implicitly provided by the Wikipedia link. For bell polynomials, the Wikipedia link is https://en.wikipedia.org/wiki/Bell_polynomials, which can be (with a quite expensive API call) rewritten to https://www.wikidata.org/wiki/Q815666. From there, you get the labels and descriptions in machine-readable forms, e.g., in JSON. To demonstrate the size, here is an example of `fraction`
```json
        "zh": {
          "language": "zh",
          "value": "分數"
        },
        "ky": {
          "language": "ky",
          "value": "алым"
        },
        "jv": {
          "language": "jv",
          "value": "wilangan pecahan"
        },
        "eu": {
          "language": "eu",
          "value": "zatiki"
        },
        "pl": {
          "language": "pl",
          "value": "ułamek"
        },
        "qu": {
          "language": "qu",
          "value": "ch'iqtaku"
        },
        "gd": {
          "language": "gd",
          "value": "bloigh"
        },
        "es": {
          "language": "es",
          "value": "fracción"
        },
        "ta": {
          "language": "ta",
          "value": "பின்னம்"
        },
        "ms": {
          "language": "ms",
          "value": "pecahan"
        },
        "bn": {
          "language": "bn",
          "value": "ভগ্নাংশ"
        },
        "sq": {
          "language": "sq",
          "value": "thyesa"
        },
        "el": {
          "language": "el",
          "value": "κλάσμα"
        },
        "sv": {
          "language": "sv",
          "value": "bråk"
        },
        "nl": {
          "language": "nl",
          "value": "breuk"
        },
        "ar": {
          "language": "ar",
          "value": "كسر"
        },
        "pt": {
          "language": "pt",
          "value": "fração"
        },
        "is": {
          "language": "is",
          "value": "almennt brot"
        },
        "eo": {
          "language": "eo",
          "value": "frakcio"
        },
        "yi": {
          "language": "yi",
          "value": "בראכטייל"
        },
        "ru": {
          "language": "ru",
          "value": "дробь"
        },
        "sr": {
          "language": "sr",
          "value": "разломак"
        },
        "ckb": {
          "language": "ckb",
          "value": "کەرت"
        },
        "tr": {
          "language": "tr",
          "value": "kesir"
        },
        "sn": {
          "language": "sn",
          "value": "dimbu"
        },
        "mk": {
          "language": "mk",
          "value": "дропка"
        },
        "fi": {
          "language": "fi",
          "value": "murtoluku"
        },
        "uk": {
          "language": "uk",
          "value": "дріб"
        },
        "be-tarask": {
          "language": "be-tarask",
          "value": "дроб"
        },
        "gan": {
          "language": "gan",
          "value": "分數"
        },
        "nn": {
          "language": "nn",
          "value": "brøk"
        },
        "io": {
          "language": "io",
          "value": "fraciono"
        },
        "lmo": {
          "language": "lmo",
          "value": "frazziun"
        },
        "tl": {
          "language": "tl",
          "value": "praksiyon"
        },
        "da": {
          "language": "da",
          "value": "brøk"
        },
        "ko": {
          "language": "ko",
          "value": "분수"
        },
        "fr": {
          "language": "fr",
          "value": "fraction"
        },
        "be": {
          "language": "be",
          "value": "дроб"
        },
        "he": {
          "language": "he",
          "value": "שבר"
        },
        "lv": {
          "language": "lv",
          "value": "daļskaitlis"
        },
        "it": {
          "language": "it",
          "value": "frazione"
        },
        "gl": {
          "language": "gl",
          "value": "fracción"
        },
        "ay": {
          "language": "ay",
          "value": "pachjta"
        },
        "id": {
          "language": "id",
          "value": "pecahan"
        },
        "de": {
          "language": "de",
          "value": "Bruch"
        },
        "ja": {
          "language": "ja",
          "value": "分数"
        },
        "ml": {
          "language": "ml",
          "value": "ഭിന്നസംഖ്യ"
        },
        "vi": {
          "language": "vi",
          "value": "phân số"
        },
        "yue": {
          "language": "yue",
          "value": "分數"
        },
        "en": {
          "language": "en",
          "value": "fraction"
        },
        "scn": {
          "language": "scn",
          "value": "frazzioni"
        },
        "hi": {
          "language": "hi",
          "value": "भिन्न"
        },
        "sk": {
          "language": "sk",
          "value": "zlomok"
        },
        "th": {
          "language": "th",
          "value": "เศษส่วน"
        },
        "ca": {
          "language": "ca",
          "value": "fracció"
        },
        "mg": {
          "language": "mg",
          "value": "ampaha"
        },
        "sl": {
          "language": "sl",
          "value": "ulomek"
        },
        "cs": {
          "language": "cs",
          "value": "zlomek"
        },
        "bg": {
          "language": "bg",
          "value": "дроб"
        },
        "fa": {
          "language": "fa",
          "value": "کسر"
        },
        "nds": {
          "language": "nds",
          "value": "Bröök"
        },
        "lt": {
          "language": "lt",
          "value": "trupmena"
        },
        "nb": {
          "language": "nb",
          "value": "brøk"
        },
        "uz": {
          "language": "uz",
          "value": "kasr"
        },
        "as": {
          "language": "as",
          "value": "ভগ্নাংশ"
        },
        "az": {
          "language": "az",
          "value": "kəsr"
        },
        "my": {
          "language": "my",
          "value": "အပိုင်းဂဏန်း"
        },
        "sr-ec": {
          "language": "sr-ec",
          "value": "разломак"
        },
        "oc": {
          "language": "oc",
          "value": "fraction"
        },
        "hr": {
          "language": "hr",
          "value": "razlomak"
        },
        "et": {
          "language": "et",
          "value": "murdarv"
        },
        "pt-br": {
          "language": "pt-br",
          "value": "fração"
        },
        "sco": {
          "language": "sco",
          "value": "fraction"
        },
        "br": {
          "language": "br",
          "value": "rann"
        },
        "la": {
          "language": "la",
          "value": "fractio"
        },
        "ro": {
          "language": "ro",
          "value": "fracție"
        },
        "af": {
          "language": "af",
          "value": "breuk"
        },
        "te": {
          "language": "te",
          "value": "భిన్నాలు"
        },
        "sh": {
          "language": "sh",
          "value": "razlomak"
        },
        "xh": {
          "language": "xh",
          "value": "i-fraction"
        },
        "cy": {
          "language": "cy",
          "value": "ffracsiwn"
        },
        "en-ca": {
          "language": "en-ca",
          "value": "fraction"
        },
        "en-gb": {
          "language": "en-gb",
          "value": "fraction"
        },
        "ba": {
          "language": "ba",
          "value": "кәсерҙәр"
        },
        "jam": {
          "language": "jam",
          "value": "frakshan"
        },
        "lg": {
          "language": "lg",
          "value": "omukutule ogwa bulijjo"
        },
        "ur": {
          "language": "ur",
          "value": "کسر"
        },
        "lfn": {
          "language": "lfn",
          "value": "frato"
        },
        "hy": {
          "language": "hy",
          "value": "կոտորակ"
        },
        "ast": {
          "language": "ast",
          "value": "fraición"
        },
        "kn": {
          "language": "kn",
          "value": "ಭಿನ್ನಾಂಕ"
        },
        "tk": {
          "language": "tk",
          "value": "drob"
        },
        "cv": {
          "language": "cv",
          "value": "вак"
        },
        "hu": {
          "language": "hu",
          "value": "törtszám"
        },
        "gcr": {
          "language": "gcr",
          "value": "fraksyon"
        },
        "wuu": {
          "language": "wuu",
          "value": "分数"
        },
        "zh-hant": {
          "language": "zh-hant",
          "value": "分數"
        },
        "zh-hk": {
          "language": "zh-hk",
          "value": "分數"
        },
        "frr": {
          "language": "frr",
          "value": "Brööktaal"
        },
        "ug": {
          "language": "ug",
          "value": "كەسىر"
        },
        "bcl": {
          "language": "bcl",
          "value": "praksyon"
        },
        "war": {
          "language": "war",
          "value": "fraksyon"
        },
        "mi": {
          "language": "mi",
          "value": "hautau"
        },
        "blk": {
          "language": "blk",
          "value": "ဂဏန်ꩻခေါဝ်"
        },
        "bs": {
          "language": "bs",
          "value": "razlomak"
        },
        "om": {
          "language": "om",
          "value": "hirmeemmii"
        },
        "mt": {
          "language": "mt",
          "value": "frazzjoni"
        },
        "kab": {
          "language": "kab",
          "value": "Tirẓi (tusnakt)"
        }
```
So, it might even make sense to store the translations in separate files. 

> In core we have got away with informally describing the argument meaning for all arguments together as part of a single comment. Are you suggesting that the yaml should have specific field for each argument that describes it?

It would make the most sense if I imported the core first. It is likely more convenient for human translators to have that as a single comment. But I think the effort for to create lists of argument names is lower. So either way is fine. 

I don't want to hold back this PR. Thus, I guess optional argument names are a topic for a future PR (if we see a perspective that those are used).

However, changing the description field to en would be wrong if we decided to go with one translation file per language.